nefertari/scripts/es.py

```diff
@@ -1,138 +1,272 @@
+    log.info('Creating mappings')
+    ES.setup_mappings()
 
-        if not self.options.quiet:
-            self.log.setLevel(logging.INFO)
 
-        self.settings = dictset(registry.settings)
+class ESTasksPool(object):
 
-    def index_models(self, model_names):
-        self.log.info('Indexing models documents')
-        params = self.options.params or ''
-        params = dict([[k, v[0]] for k, v in urllib.parse.parse_qs(params).items()])
+    def __init__(self, model_names, processes, options):
+        self._pool = {process_id: [] for process_id in range(0, processes)}
+        self.options = options
 
         for model_name in model_names:
-            self.log.info('Processing model `{}`'.format(model_name))
+            for process_id in range(0, processes):
+                self._pool[process_id].append(ESTask(model_name, options))
+
+    def get_task(self, process_id):
+        if not self.is_empty(process_id):
+            return self._pool[process_id].pop()
+        return None
+
+
+    def is_empty(self, process_id):
+        return len(self._pool[process_id]) == 0
+
+
+class ESProcessor(object):
+
+    def __init__(self, options, pool, process_id, model_names, processes_count):
+        self.options = options
+        self.pool = pool
+        self.id = process_id
+        self.processes_count = processes_count
+        self.model_names = model_names
+
+    @staticmethod
+    def split_collection(limit, n, collection):
+        list_size = (math.ceil(limit / n) - 1) or 1
+        iteration = 0
+
+        for i in range(0, limit, list_size):
+            iteration += 1
+
+            if iteration == n:
+                yield collection[i:]
+                break
+
+            yield collection[i: i + list_size]
+
+    def get_chunks(self):
+        model_chunks = {}
+        from nefertari import engine
+        from pyramid_sqlalchemy import Session
+
+        for model_name in self.model_names:
             model = engine.get_document_cls(model_name)
+            limit = model.get_collection().count()
+            table_name = model.__tablename__
+            statement = text('SELECT {} FROM public.{}'.format(model.pk_field(), table_name))
+            query = Session().query(model).from_statement(statement)
+            items = list(sorted(query.values(model.pk_field())))
+            chunks = list(self.split_collection(limit, self.processes_count, items))
 
-            local_params = dict()
-            local_params.update(params)
+            if len(chunks) - 1 < self.id:
+                model_chunks[model_name] = tuple()
+                continue
 
-            if '_limit' not in local_params:
-                limit = model.get_collection().count()
-                local_params['_limit'] = limit
+            model_chunks[model_name] = chunks[self.id]
+        return model_chunks
 
-            chunk_size = int(self.options.chunk or local_params['_limit'])
+    def __call__(self):
```

Comment:
This is wrapper, because we can't pass object method as high order function to `map`